bftengine/tests/controllerWithSimpleHistory/ControllerWithSimpleHistory_test.cpp

```diff
@@ -145,7 +147,319 @@ TEST(ControllerWithSimpleHistory, onNewView_not_primary) {
   ASSERT_EQ(true, cwsh.insideActiveWindow(((sn - 1) + ControllerWithSimpleHistory::EvaluationPeriod)));
 }
 
-int main(int argc, char** argv) {
+////////////////////////////Path upgrade/downgrade scenarios/////////////////////
+
+/// Downgrade scenarios
+
+// Test - Downgrade from OPTIMISTIC_FAST to FAST_WITH_THRESHOLD
+// Logic:
+// 1)Construct controller with C > 0 i.e. enable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call onStartingSlowCommit for more than factor * EvaluationPeriod.
+TEST(ControllerWithSimpleHistory, downgrade_from_optimistic_to_threshold) {
+  uint16_t C = 1;
+  uint16_t F = 1;
+  ReplicaId replicaId = 0;
+  ViewNum initialView = 0;
+  SeqNum initialSeq = 0;
+  ControllerWithSimpleHistory cwsh{C, F, replicaId, initialView, initialSeq};
+
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+
+  // Factor + 1, in order to trigger the degradation.
+  auto slowStartCount =
+      ControllerWithSimpleHistory_debugDowngradeFactor * ControllerWithSimpleHistory::EvaluationPeriod + 1;
+  auto prePrepareTime = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+  bool changed{false};
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    if (slowStartCount-- > 0) cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+  ASSERT_EQ(changed, true);
+  ASSERT_EQ(CommitPath::FAST_WITH_THRESHOLD, cwsh.getCurrentFirstPath());
+}
+
+// Test - Downgrade from OPTIMISTIC_FAST to SLOW
+// Logic:
+// 1)Construct controller with C == 0 i.e. disable FAST_WITH_THRESHOLD path.
+// 2)Run loop for EvaluationPeriod times:
+//  - call onStartingSlowCommit for more than factor * EvaluationPeriod.
+TEST(ControllerWithSimpleHistory, downgrade_from_optimistic_to_slow) {
+  uint16_t C = 0;
+  uint16_t F = 1;
+  ReplicaId replicaId = 0;
+  ViewNum initialView = 0;
+  SeqNum initialSeq = 0;
+  ControllerWithSimpleHistory cwsh{C, F, replicaId, initialView, initialSeq};
+
+  ASSERT_EQ(CommitPath::OPTIMISTIC_FAST, cwsh.getCurrentFirstPath());
+
+  // Factor + 1, in order to trigger the degradation.
+  auto slowStartCount =
+      ControllerWithSimpleHistory_debugDowngradeFactor * ControllerWithSimpleHistory::EvaluationPeriod + 1;
+  auto prePrepareTime = std::chrono::steady_clock::now() - std::chrono::milliseconds(10);
+  bool changed{false};
+  for (auto i = (size_t)1; i <= ControllerWithSimpleHistory::EvaluationPeriod; ++i) {
+    cwsh.onSendingPrePrepare((SeqNum)i, CommitPath::OPTIMISTIC_FAST, prePrepareTime);
+    changed = cwsh.onNewSeqNumberExecution((SeqNum)i);
+    if (slowStartCount-- > 0) cwsh.onStartingSlowCommit((SeqNum)i);
+  }
+  ASSERT_EQ(changed, true);
+  ASSERT_EQ(CommitPath::SLOW, cwsh.getCurrentFirstPath());
```

Comment:
So, if we consider this at the level of a live blockchain/BFT network, is it true that:

if C=0 and a replica crashes, `getCurrentFirstPath()` returns OPTIMISTIC_FAST before the evaluation period, and SLOW right afterwards?

Reply:
Yes, but in that case there is another mechanism, that detects active requests, which didn't complete the fast path during a given interval. It downgrades the detected requests to the slow path.
`ReplicaImp::tryToStartSlowPaths`.